### DISCUSSION TITLE
Add an option to display tokens/s in real-time for both side-by-side and single model inference.

### DISCUSSION THREAD

**Discussion by zytoh0** [open-webui/open-webui#5455] [General]
**Is your feature request related to a problem? Please describe.**
Open WebUI’s comparison feature is useful for evaluating output quality, but it lacks the ability to show tokens per second (tokens/s), which is important for comparing model speed. This is missing in both side-by-side and single model inference. UIs like Jan.ai already include this feature, making it easier to assess model performance. (See image below). Visit https://www.youtube.com/watch?v=QpMQgJL4AZA and scroll to 02:48 to see it display token/s in real-time.
![image](https://github.com/user-attachments/assets/fb3be782-f2ca-4cf1-aacd-2d03fa8e5482)

**Describe the solution you'd like**
Add a setting that displays tokens/s in real-time for both side-by-side comparisons and single model inferences. This would allow users to track model speed alongside output quality.

**Describe alternatives you've considered**
Another option could be to display tokens/s within the response itself, either below or next to the text for every single response but having a setting to toggle this feature on or off would be more flexible for users who may not always need this information.

**Additional context**
Adding this feature would greatly improve the usability of Open WebUI by allowing users to evaluate both model output quality and processing speed. This is especially useful when speed is a critical factor.

**Comment by velum** [comment]
That would be a very useful feature, and I am wondering why this discussion is marked as "Closed".

**Reply by zytoh0** [reply]
Cause I've replicated under ideas here: https://github.com/open-webui/open-webui/discussions/5451
Seems this has more upvotes than the replicated one, as such I've reopened it :)

**Comment by zytoh0** [comment]
> That would be a very useful feature, and I am wondering why this discussion is marked as "Closed".

They've implemented this feature. Closing this as resolved.

**Reply by orrinwitt** [reply]
Could you tell us where in the settings we can find this feature? I'm having a hard time finding it.

**Reply by lordgabor** [reply]
> Could you tell us where in the settings we can find this feature? I'm having a hard time finding it.

In the generated answer, hover over the (i) icon (the fourth one right now).
![InfoButton](https://github.com/user-attachments/assets/ac31b414-feb2-42b4-8d58-8f404d29bb8d)

**Reply by cwallace** [reply]
That icon doesn't exist when I am using external models:

![image](https://github.com/user-attachments/assets/b6e50e50-95af-40f3-b005-3515e6a98110)

**Reply by patdmob** [reply]
Even for other locally hosted models (ie LM Studio has different models available and I wanted to test some of them) and they aren't showing because they go through the "OpenAI API"... :/

**Reply by cloudliu01** [reply]
Any particular reason it's not available for 'OpenAI API' users? Thanks!

**Comment by JamesMowery** [comment]
@zytoh0 requesting that you reopen this.

I'm running models locally with Llama.cpp and I only see `completion_tokens`, `prompt_tokens`, and `total_tokens` under the (i) hover. I do not have tokens per second, it's nowhere to be found. So this is definitely not working or there is a bug.

**Reply by zytoh0** [reply]
@JamesMowery It only works with ollama models. I've reopened. Hopefully, they will support other inference servers.

**Reply by jhgoodwin** [reply]
@zytoh0 Do you have the issue number or link?

**Reply by o0LINNY0o** [reply]
same setup @zytoh0 , is there any future planned implementation

**Comment by vigneshwarrvenkat** [comment]
Hi everyone,

I’m adding better speed and usage info for replies, and I want to match what this thread already asked for so I don’t duplicate work.

What we have today
When a reply finishes, the small ⓘ button can show usage from the model server (token counts and, for Ollama, timing info inside that same block). That’s already useful.

What I want to add
I want to build on that same ⓘ button, not replace it:

Simple numbers from the browser — e.g. how long until the first words appear, how long the full reply took, and a rough “tokens per second” when the server tells us how many tokens were completed.
Still show the server’s usage in the tooltip, like today.
One click on ⓘ copies a small JSON file people can paste in issues or when comparing models.
Optional extra panel (“Observe”) for people who want a list of recent runs, loaded Ollama models, or token probability details when the API supports it. This is extra, the main place for metrics stays ⓘ on each reply.
What I’m not trying to do
I’m not replacing Arena or the leaderboard here.

Question for the team
Is this the right direction for #5455 before I open a PR to the dev branch?
If something is already being built the same way, please point me to it so I can align or drop overlap.

**Reply by Classic298** [reply]
why not do it with a filter?

**Comment by Classic298** [comment]
Why not implement this as a filter?

**Reply by vigneshwarrvenkat** [reply]
@Classic298 Thanks for the reply
when you say “as a filter,” do you mean a Pipelines filter plugin, or a UI/settings toggle to show vs hide metrics?
If it’s Pipelines: happy to expose extra server-side hooks, but time-to-first-token and per-message ⓘ need browser/UI wiring, so core (or another frontend contribution) still has a role unless we scoped this to logging only.
If it’s UI filtering: yes — we can keep metrics optional behind settings so normal users aren’t cluttered. Happy to align with whatever pattern is best.

**Reply by Classic298** [reply]
I mean a filter function. Not pipeline.

**Reply by vigneshwarrvenkat** [reply]
When you say implement this as a filter function, do you mean everything should live in a Python filter (inlet/outlet/stream) and we should avoid changes to the web UI (no ⓘ / no panel), or is a small UI change still OK to display what the filter adds?

**Reply by Classic298** [reply]
the filter modifies the UI through event emitter

**Reply by vigneshwarrvenkat** [reply]
Got it. You mean filters can attach an event path so event_emitter reaches the UI, not that everything stays invisible on the backend.
Thanks, that clarifies direction: implement observability signals from a stream filter via event, and only add frontend where we need to handle a new payload type.
Is there an existing event type pattern you prefer for lightweight metrics (or should we align with chat:completion data extras)?

**Reply by Classic298** [reply]
dont modify the frontend directly as I said - use the filter with the correct event emitter

**Reply by vigneshwarrvenkat** [reply]
Understood. I’ll try to keep this in a stream filter and drive the UI only through event_emitter using existing types (chat:completion for usage / deltas, and status for readable metric lines). I won’t add new frontend code for this path.

**Reply by vigneshwarrvenkat** [reply]
@Classic298 I raised a PR. Can you pls take a look.
https://github.com/open-webui/open-webui/pull/24528

**Reply by vigneshwarrvenkat** [reply]
@Classic298 With regards to the comment in https://github.com/open-webui/open-webui/pull/24528
I have few clarification. 
Are you recommending the changes to this repo https://github.com/open-webui/docs
Can you please clarify?

**Reply by Classic298** [reply]
yes. I like your filter it works GREAT.

You did not implement it the way i thought you would - but it does the job

I will review your PR and give feedback

for now already: code should be more minimal to be merged into docs PR and be inline in details block not an extra file

**Reply by vigneshwarrvenkat** [reply]
@Classic298 I have raised this PR. https://github.com/open-webui/docs/pull/1250 
Please provide your feedback.

**Reply by Classic298** [reply]
@vigneshwarrvenkat for the docs or for the filter?

for docs: the filter is a bit too long to merge it into the docs - but I would happily accept your textual changes. Feel free to expand them into describing what such a filter does/would do and behaves.

Did you publish your filter to openwebui.com yet ?

**Reply by vigneshwarrvenkat** [reply]
@Classic298 Made the changes to the PR https://github.com/open-webui/docs/pull/1250
url : https://openwebui.com/posts/461c301e-30c4-46d1-a8aa-eb5ece7be85f

Please provide your feedback.

**Reply by vigneshwarrvenkat** [reply]
@Classic298 Did you get a chance to look at the PR. Any feedback on the PR.
Please let me know.

**Reply by vigneshwarrvenkat** [reply]
@Classic298 Thanks for the DEV merge.
Any idea when this will be merged to the main and be available for usage.
